src/jupyterlabcontroller/services/image.py

```diff
@@ -430,6 +345,39 @@ async def prepuller_wait(self) -> None:
         await self._refreshed.wait()
         self._refreshed.clear()
 
+    def _build_node_images(
+        self,
+        to_prepull: RSPImageCollection,
+        node_cache: dict[str, list[KubernetesNodeImage]],
+    ) -> dict[str, RSPImageCollection]:
+        """Construct the collection of images on each node.
+
+        Parameters
+        ----------
+        to_prepull
+            Images that should be prepulled, and against which we compare
+            cached images.
+        node_cache
+            Mapping of node names to the list of cached images on that node.
+
+        Returns
+        -------
+        dict of str to RSPImageCollection
+            Image collections of images found on each node.
+        """
+        image_collections = {}
+        for node, node_images in node_cache.items():
```

Comment:
If you're into abusing comprehensions, you could do this as a nested pair of list comprehensions.  Please don't though.